composer/datasets/in_context_learning_evaluation.py

```diff
@@ -85,582 +109,764 @@ def _make_padded_input(context_enc, continuation_enc, max_seq_len, pad_tok_id, p
+        self.passage_delimiter = passage_delimiter
+        self.passage_query_delimiter = passage_query_delimiter
+        super().__init__(*args, **kwargs)
+
+    def _construct_context(self, example: dict, preceding_text: str = '', add_answer: bool = False):
+        """
+        Takes a example and constructs a context. Optionally, appends this to preceeding text (such as a
+        prompt or fewshot examples), as well as optionally adds the correct answer (for fewshot examples)
 
-        self.prefix_space = _tokenizer_needs_prefix_space(self.tokenizer)
+        Args:
+            example (dict): the example from which to construct the context
+            preceding_text (str): any preceding text, needed to if self.example_delimiter is needed at the beginning
+            add_answer (bool): bool for whether or not to add the answer on the end of the context (needed for fewshot examples)
 
-        self.encoded_dataset = self.prep_examples(num_fewshot, prompt_string, example_delimiter, continuation_delimiter,
-                                                  fewshot_rng)
+        Returns:
+            str: The constructed context. The default output context is
+                 formatted as follows: f'{self.prelimiter}{example['self.passages_key']}{example[self.context_key]}{self.continuation_delimiter}'
+        """
+        passages = self.passage_delimiter.lstrip('\n ')
+        passages += f'{self.passage_delimiter}'.join(example['passages'])
+        query = example['query']
+        # TODO: add few_shot capabilities
+        context = f'{self.prelimiter}{passages}{self.passage_query_delimiter}{query}'
+        return context
+
+    def _tokenize_example(self, prompt_and_fewshot: str, ctxt: str, example: dict):
+        """
+        Runs text through the tokenizer and handles special cases.
+        Args:
+            prompt_and_fewshot (str): the collection of the prompt and fewshot examples that belongs before the example's context
+            ctx (str): the specific example's derived context
+            example (dict): the example as a dictionary.
 
-    def prep_examples(self, num_fewshot: int, prompt_string: str, example_delimiter: str, continuation_delimiter: str,
-                      fewshot_rng: random.Random):
-        """Prepares a set of language modeling tasks into tokenized format with prompt and fewshot examples.
+        Returns:
+            dict: dictionary with the tokenized data
+        """
+        tokenized_example = super()._tokenize_example(prompt_and_fewshot, ctxt, example)
+        answer = example['answers'][0]
+        tokenized_example['answer'] = self.tokenizer(answer, add_special_tokens=False)
+        return tokenized_example
 
-        Each task consists of a context and a continuation as well as an optional prompt and optional list of
-        example context/continuation pairs which precede the test context/continuation pair.
 
+    def collate_fn(self, data):
```

Comment:
It looks like this collate_fn is set up to do something resembling a LM task (i.e. call `forward`) should it instead be set up as a QA task (i.e. call `generate`)?

Reply:
This is Eitan's RAG task. It's WIP. It shouldn't be on this branch but he pushed it. I'll take it out and put it on it's own branch